python/lsst/daf/butler/queries/_base.py

```diff
@@ -0,0 +1,287 @@
+
+    @property
+    def dimensions(self) -> DimensionGroup:
+        """All dimensions included in the query's columns."""
+        return self._tree.dimensions
+
+    def any(self, *, execute: bool = True, exact: bool = True) -> bool:
+        # Docstring inherited.
+        return self._driver.any(self._tree, execute=execute, exact=exact)
+
+    def explain_no_results(self, execute: bool = True) -> Iterable[str]:
+        # Docstring inherited.
+        return self._driver.explain_no_results(self._tree, execute=execute)
+
+
+class CountableQueryBase(QueryBase):
+    """Common base class for query result objects for which the number of
+    result rows is a well-defined concept.
+
+    This class should rarely be referenced directly; it is public only because
+    it provides public methods to its subclasses.
+    """
+
+    @abstractmethod
+    def count(self, *, exact: bool = True, discard: bool = False) -> int:
+        """Count the number of rows this query would return.
+
+        Parameters
+        ----------
+        exact : `bool`, optional
+            If `True`, run the full query and perform post-query filtering if
+            needed to account for that filtering in the count.  If `False`, the
+            result may be an upper bound.
+        discard : `bool`, optional
+            If `True`, compute the exact count even if it would require running
+            the full query and then throwing away the result rows after
+            counting them.  If `False`, this is an error, as the user would
+            usually be better off executing the query first to fetch its rows
+            into a new query (or passing ``exact=False``).  Ignored if
+            ``exact=False``.
+
+        Returns
+        -------
+        count : `int`
+            The number of rows the query would return, or an upper bound if
+            ``exact=False``.
+        """
+        raise NotImplementedError()
+
+
+class QueryResultsBase(HomogeneousQueryBase, CountableQueryBase):
+    """Common base class for query result objects with homogeneous dimensions
+    and countable rows.
+    """
+
+    def order_by(self, *args: str | OrderExpression | ExpressionProxy) -> Self:
+        """Return a new query that yields ordered results.
+
+        Parameters
+        ----------
+        *args : `str`
+            Names of the columns/dimensions to use for ordering. Column name
+            can be prefixed with minus (``-``) to use descending ordering.
+
+        Returns
+        -------
+        result : `QueryResultsBase`
+            An ordered version of this query results object.
+
+        Notes
+        -----
+        If this method is called multiple times, the new sort terms replace
+        the old ones.
+        """
+        return self._copy(
+            self._tree, order_by=convert_order_by_args(self.dimensions, self._get_datasets(), *args)
+        )
+
+    def limit(self, limit: int | None = None, offset: int = 0) -> Self:
```

Comment:
I think the intent was indeed to support paging. Efficiency of course depends on implementation. For complex queries it may be very inefficient, but for materialized tables it may work fine. OK for me to drop it if we don't use it.